business/authentication/token_auth_controller.go

```diff
@@ -0,0 +1,214 @@
+	// Get the token from HTTP form data
+	err := r.ParseForm()
+	if err != nil {
+		return nil, fmt.Errorf("error parsing form data from client: %w", err)
+	}
+
+	token := r.PostForm.Get("token")
+	if token == "" {
+		return nil, errors.New("token is empty")
+	}
+
+	// Create a bs layer with the received token to check its validity.
+	bs, err := c.businessInstantiator(&api.AuthInfo{Token: token})
+	if err != nil {
+		return nil, fmt.Errorf("error instantiating the business layer: %w", err)
+	}
+
+	// Using the namespaces API to check if token is valid. In Kubernetes, the version API seems to allow
+	// anonymous access, so it's not feasible to use the version API for token verification.
+	nsList, err := bs.Namespace.GetNamespaces()
+	if err != nil {
+		c.SessionStore.TerminateSession(r, w)
+		return nil, &AuthenticationFailureError{Reason: "token is not valid or is expired", Detail: err}
+	}
+
+	// If namespace list is empty, return authentication failure.
+	if len(nsList) == 0 {
+		c.SessionStore.TerminateSession(r, w)
+		return nil, &AuthenticationFailureError{Reason: "not enough privileges to login"}
+	}
+
+	// Token was valid against the Kubernetes API, and it has privileges to read some namespace.
+	// Accept the token. Create the user session.
+	timeExpire := util.Clock.Now().Add(time.Second * time.Duration(config.Get().LoginToken.ExpirationSeconds))
+	err = c.SessionStore.CreateSession(r, w, "token", timeExpire, sessionPayload{Token: token})
+	if err != nil {
+		return nil, err
+	}
+
+	return &UserSessionData{ExpiresOn: timeExpire, Username: extractSubjectFromK8sToken(token), Token: token}, nil
+}
+
+// ValidateSession restores a session previously created by the Authenticate function. A minimal re-validation
+// is done: only token validity is re-checked by making a request to the Kubernetes API, like in the Authenticate
+// function. However, privileges are not re-checked.
+// If the session is still valid, a populated UserSessionData is returned. Otherwise, nil is returned.
+func (c tokenAuthController) ValidateSession(r *http.Request, w http.ResponseWriter) (*UserSessionData, error) {
+	// Restore a previously started session.
+	sPayload := sessionPayload{}
+	sData, err := c.SessionStore.ReadSession(r, w, &sPayload)
+	if err != nil {
+		log.Warningf("Could not read the session: %v", err)
+		return nil, nil
+	}
+	if sData == nil {
+		return nil, nil
+	}
+
+	// Check token validity.
+	bs, err := c.businessInstantiator(&api.AuthInfo{Token: sPayload.Token})
+	if err != nil {
+		return nil, fmt.Errorf("could not get the business layer: %w", err)
+	}
+
+	_, err = bs.Namespace.GetNamespaces()
+	if err == nil {
+		// Internal header used to propagate the subject of the request for audit purposes
+		r.Header.Add("Kiali-User", extractSubjectFromK8sToken(sPayload.Token))
+		return &UserSessionData{ExpiresOn: sData.ExpiresOn, Username: extractSubjectFromK8sToken(sPayload.Token), Token: sPayload.Token}, nil
+	}
```

Comment:
nit: can this be moved to the end of the function and the check be changed to `err != nil` as the others above it?

Reply:
Yes